File: src/main/java/com/skysoft/utils/animation/AnimationClock.java
package com.skysoft.utils.animation;

import java.util.function.LongSupplier;

public final class AnimationClock {
    private final LongSupplier nanoTime;
    private long startedAtNanos = Long.MIN_VALUE;

    public AnimationClock() {
        this(System::nanoTime);
    }

    AnimationClock(LongSupplier nanoTime) {
        this.nanoTime = nanoTime;
    }

    public void restart() {
        startedAtNanos = nanoTime.getAsLong();
    }

    public void stop() {
        startedAtNanos = Long.MIN_VALUE;
    }

    public boolean hasStarted() {
        return startedAtNanos != Long.MIN_VALUE;
    }

    public float progress(int durationMillis) {
        if (!hasStarted()) {
            return 1.0f;
        }
        if (durationMillis <= 0) {
            return 1.0f;
        }

        long durationNanos = durationMillis * 1_000_000L;
        long elapsedNanos = Math.max(0L, nanoTime.getAsLong() - startedAtNanos);
        return Math.min(elapsedNanos / (float) durationNanos, 1.0f);
    }
}

File: src/main/kotlin/com/skysoft/utils/animation/AnimationClock.kt
package com.skysoft.utils.animation

import java.util.function.LongSupplier

class AnimationClock internal constructor(
    private val nanoTime: LongSupplier,
) {
    constructor() : this(LongSupplier(System::nanoTime))

    private var startedAtNanos = Long.MIN_VALUE

    fun restart() {
        startedAtNanos = nanoTime.getAsLong()
    }

    fun stop() {
        startedAtNanos = Long.MIN_VALUE
    }

    fun hasStarted(): Boolean = startedAtNanos != Long.MIN_VALUE

    fun progress(durationMillis: Int): Float {
        if (!hasStarted() || durationMillis <= 0) return 1.0f
        val durationNanos = durationMillis.toLong() * NANOS_PER_MILLISECOND
        val elapsedNanos = (nanoTime.getAsLong() - startedAtNanos).coerceAtLeast(0L)
        return (elapsedNanos / durationNanos.toFloat()).coerceAtMost(1.0f)
    }

    private companion object {
        const val NANOS_PER_MILLISECOND = 1_000_000L
    }
}
